packages/flame/lib/src/components/mixins/has_paint.dart

```diff
@@ -3,19 +3,46 @@ import 'dart:ui';
 import 'package:flame/components.dart';
 import 'package:flame/src/palette.dart';
 
-/// Adds a collection of paints to a component
+/// Adds a collection of paints and paint layers to a component
 ///
 /// Component will always have a main Paint that can be accessed
 /// by the [paint] attribute and other paints can be manipulated/accessed
 /// using [getPaint], [setPaint] and [deletePaint] by a paintId of generic type
-/// [T], that can be omitted if the component only have one paint.
+/// [T], that can be omitted if the component only has one paint.
+/// [paintLayers] paints should be drawn in list order during the render. The
+/// main Paint is the first element.
 mixin HasPaint<T extends Object> on Component {
-  final Map<T, Paint> _paints = {};
+  late final Map<T, Paint> _paints = {};
 
-  Paint paint = BasicPalette.white.paint();
+  List<Paint>? _paintLayers;
 
-  void _assertGenerics() {
-    assert(T != Object, 'A generics type is missing on the HasPaint mixin');
+  Paint _paint = BasicPalette.white.paint();
+
+  /// List of paints to use (in order) during render.
+  List<Paint> get paintLayers {
+    if ((_paintLayers == null) || _paintLayers!.isEmpty) {
+      return _paintLayers = [_paint];
+    }
+    return _paintLayers!;
+  }
+
+  set paintLayers(List<Paint> paintLayers) {
+    _paintLayers = paintLayers;
+  }
+
+  /// Main paint. The first paint in the [paintLayers] list.
+  Paint get paint {
+    if ((_paintLayers == null) || _paintLayers!.isEmpty) {
```

Comment:
This is where it gets confusing to me. Why do the layers have precedence over the paint map?

On the other hand, I also can't argue why would the map have over the layers.

I wonder if we shouldn't consider layers just a different thing from the paint layers, and let the components decide how they will use?

Reply:
The paint map is independent of the main paint, and is only used for paint collections? As for the null and isEmpty checks, it's to prevent a get on paintLayers if it's not been set yet as there's no need to allocate the List for it if not used by the component itself, saving memory. Otherwise if it's not empty then paintLayers[0] holds the main paint. This is because when iterating through paintLayers during a component's `render` it avoids needing to duplicate drawing code specially for the special case of the first layer, paint.

I'm confused about the meaning of considering layers a different thing from paint layers? In HasPaint there is only one layer concept, paintLayers?

Reply:
> This is where it gets confusing to me. Why do the layers have precedence over the paint map?

The paint map doesn't really have any natural first item? But I was thinking about that too, that maybe the basic paint shouldn't be included in the paint layers at all. So that there basically are three completely separate features in here; the normal paint, the paint map and the paint layers.

> I wonder if we shouldn't consider layers just a different thing from the paint layers, and let the components decide how they will use?

The components now decide how they will use the paint layers.
What do you mean by "consider layers just a different thing from the paint layers"?

Reply:
> > This is where it gets confusing to me. Why do the layers have precedence over the paint map?
> 
> The paint map doesn't really have any natural first item? But I was thinking about that too, that maybe the basic paint shouldn't be included in the paint layers at all. So that there basically are three completely separate features in here; the normal paint, the paint map and the paint layers.

The reason is to save duplication of drawing code in components that use paint layers in their render. Otherwise they'll have to do one set of drawing using `paint` then duplicate that drawing iterating through `paintLayers`. Making `paint` tied to the first element of `paintLayers` makes things easier downstream I think.

Reply:
Indeed the paint map doesn't have a first because this is a map, which has no order. But HasPaint does have a concept of a main paint (the one that we store on the variable, which is set without an id.

I would argue that the main paint can be considered part of the paint layers. But should the first layer be considered the main paint? That is where I think I am confused.

I think that this would be less confusing if the `get paint` remained untouched, and if we agree that the main paint can be considered a layer, then it could be included on the paintLayers getter and the components that use layers would remain working the same.

Reply:
Until `paintLayers` is used, `paint` is a completely separate property as before anyway. Once `paintLayers` are used, then I think it'll be more confusing if `paintLayers` didn't actually contain all the paint layers...

Reply:
It seems what you're suggesting @erickzanardo is already what's happening? Otherwise I'm a bit confused what the issue with the implementation is? Is the concern whether the main paint should be drawn first or last? Draw first with subsequent paint layers being drawn on top seemed the natural order to me, but do others disagree?